### ISSUE TITLE
workingSpeed retourne toujours la même valeurs

### ISSUE THREAD

**Issue by andromeda2911** [eserandour/Minitel1B_Hard#18]
Hardware: Minitel2
la fonction workingSpeed me retourne toujours la même valeurs,
0x1B3AFFFF
quelque soit la vitesse.
je suis en mode videotext

**Comment by andromeda2911** [comment]
surement un problème de parité, j'ai le meme problème avec workingMode qui me retourne
0x1B3AFF48 au lieu de 0x1B3A73

**Comment by troisiemetype** [comment]
Cela pourrait être lié aux broches par défaut du serial 2 : #19.

**Comment by andromeda2911** [comment]
Depuis je ne l'ai pas testé sur le dongle iodeo, dans mon cas j'avais fait une adaptation de cette librairie pour AgonLight2, qui fonctionne maintenant (en langage c pour ez80), comme c'est du c, c'est pratiquement le même code à part tout ce qu concerne le port série.
j'ai fait une modif dans ReadByte et WorkingSpeed.
pour ReadByte!
```
uint8_t Minitel::readByte() {
  uint8_t b = serial_read();
  return b & 0x7F;
  // Le bit de parité est à 0 si la somme des autres bits est paire
  // et à 1 si elle est impaire.
  /*bool parite = 0;
  for (int i=0; i<7; i++) {
    if (bitRead(b,i) == 1)  {
      parite = !parite;
    }
  }
  if (bitRead(b,7) == parite) {  // La transmission est bonne, on peut récupérer la donnée.
    if (bitRead(b,7) == 1) {  // Cas où le bit de parité vaut 1.
      b = b ^ 0b10000000;  // OU exclusif pour mettre le bit de parité à 0 afin de récupérer la donnée.
    }    
    return b;
  }
  else {    
    return 0xFF;  // Pour indiquer une erreur de parité.
  }*/
  
}

```
et pour WorkingSpeed:
```
int Minitel::workingSpeed() {
  int bauds = -1;
  //while (!available());  // On attend que le port soit sur écoute. on enleve les commentaires si on utilise seul minitel.changeSpeed() 
  ticker = 0;
  unsigned long duree = 0;
  unsigned long trame = 0;  // 32 bits = 4 octets
.....
....
```
  
Ce qui change c'est le commentaire du while.
On enleve les commentaires si on utilise changeSpeed seulement sans le SearchSpeed.

Attention dans ton explication tu as mis Serial2.begin(1200, SERIAL_8N1, 16, 17); alors que c'est
Serial2.begin(1200, SERIAL_7E1, 16, 17);
 c'est du 7 bit Even et 1.

Sur le principe je suis d'accord d'ajouter deux membres et de modifier le constructeur ou ajouter un constructeur avec les paramètres RX et TX.

**Comment by troisiemetype** [comment]
C'est exact, pour la vitesse de transfert. Je viens de vérifier dans les spécifications du Minitel, c'est 7N1. J'avais mis 8N1 puisque la bibliothèque ne le précise pas, et que c'est la valeur par défaut. Et comme ça fonctionne avec 8N1 dans les deux sens et à tous les baudrate, je ne me suis pas posé de question. :)

**Comment by andromeda2911** [comment]
Ca fonctionnait car dans le code, le calcul de parité est fait automatiquement, dans la fonction ReadByte.
J'en ai profité pour modifier cette bibliothèque au niveau du constructeur avex rx et tx par default, si on veut précider les pins ca devient par exemple:
```
Minitel minitel(Serial2, 16, 17); ou 
Minitel minitel(Serial2);  // Defaut c'est 16 et 17
```
et j'ai modifié tous les 
```
mySerial.begin(1200);
par
 mySerial.begin(1200, SERIAL_7E1, rx_pin, tx_pin);
```
Le constructeur dans Minitel1B_Hard.h devient:
```
Minitel(HardwareSerial& serial, uint8_t rx = 16, uint8_t tx = 17);
Et dans le fichier Minitel1B_Hard ca devient:
Minitel::Minitel(HardwareSerial& serial, uint8_t rx, uint8_t tx) : mySerial(serial), rx_pin(rx), tx_pin(tx) {
  // A la mise sous tension du Minitel, la vitesse des échanges entre
  // le Minitel et le périphérique est de 1200 bauds par défaut.
  mySerial.begin(1200, SERIAL_7E1, rx_pin, tx_pin);
}
```

**Comment by andromeda2911** [comment]
Dans arduino dans mon cas j'utilise 
Minitel minitel(Serial2, 17, 16); // 17 en premier au lieu de 16
Car j'ai ajouté un coupleur sur iodeo, femelle femelle et et j'ai connecté le cable usb din minitel sur le PC pour tester
avec hyperterminal ou autre, (donc rs et tx sont inversés)
la demo fonctionne bien , sauf pour tous ce qui est les entrées au clavier ou les demande d'infos, comme le bauds etc..
Ce qui est normal car les emulateurs ne reconnaissent pas le protocole minitel, juste pour l'affichage.
Je n'ai pas trouvé d'émulateur qui comprenne complètement le protocole minitel.
Ca m'aurais évité de sortir mon minitel qui est encombrant et lourd et faire tous mes tests juste avec un émulateur.

**Comment by troisiemetype** [comment]
En effet, je n'avais pas réalisé que le contrôle de parité était fait à la main au lieu d'utiliser les mécanismes internes du microcontrolleur.

Pour les constructeurs multiples, c'est un peu plus compliqué que prévu, puisque si l'on veut que la bibliothèque reste portable, il faudrait récupérer les valeurs par défaut. Or toutes les plateformes ne les définissent pas forcément, ni non plus de la même manière. Je n'ai pas encore trouvé de quelle manière le faire.

Pour l'émulateur... Il faudrait le coder. :) Ça pourrait se faire assez simplement avec Processing.

**Comment by andromeda2911** [comment]
Il ya un constructeur seulement que j'ai modifié, suivant les esp32, on ne connait pas à l'avance les pins de serial2, je ne sais pas si il existe sous arduino une commande qui récupère les pins par défaut, comme elles y sont déjà dans le core ESP32 dans le répertoire variant et le fichiers pins_arduino.h, quelque chose comme cela.
même sur les pico ou autre, en ces cas là, il suffit de passer en paramètres les bons pins et ca reste portable.
Suivant les platines à base de ESP32, les pins sont différents.
Sinon pour l'émulateur, en effet, il faut le créer soit même, il faut utiliser le cahier des charge du minitel.

**Comment by troisiemetype** [comment]
> Il ya un constructeur seulement que j'ai modifié, suivant les esp32, on ne connait pas à l'avance les pins de serial2, je ne sais pas si il existe sous arduino une commande qui récupère les pins par défaut, comme elles y sont déjà dans le core ESP32 dans le répertoire variant et le fichiers pins_arduino.h, quelque chose comme cela. même sur les pico ou autre, en ces cas là, il suffit de passer en paramètres les bons pins et ca reste portable. Suivant les platines à base de ESP32, les pins sont différents. Sinon pour l'émulateur, en effet, il faut le créer soit même, il faut utiliser le cahier des charge du minitel.

J'ai passé une peu de temps aujourd'hui à chercher les différentes implémentations, pour me fiare une idée de ce qu'il est possible de faire.

L'ESP 32 a sa propre méthode `begin()`, à laquelle on peut passer les numéros d'IO et d'autres options.
Le RP2040 fournit des méthodes `setRX()` et `setTX()`.
Les autre implémentations courantes (notament AVR CORE) fournissent les deux méthodes `begin()` classique.

Les architectures qui permettent de changer les broches fournissent ou bien une fonction permettant de les obtenir, ou bien des constantes, mais il n'y a aucune convention de nommage donc rien d'utilisable, à part au cas par cas.

[J'ai fait un fork](https://github.com/troisiemetype/Minitel1B) de la bibliothèque ou j'implémente la fonction d'origine par défaut **avec** **SERIAL_7E1**, et la fonction avec la selection des broches, le tout encadré par des directives preprocesseur. Voir la branche _serial_pins_.
J'ai implémenté aussi le RP2040, sans vérification de la cohérence des IO avec les capacités matérielles.

**Comment by andromeda2911** [comment]
Merci je vais regarder cela, sinon j'ai vu que ideo a fait un fork et a fait ce genre de modification pour l'
ESP32.

**Comment by troisiemetype** [comment]
Oui, en effet. On a quasiment écrit la même chose. Sur l'ESP32 il y a une fonction qui permet de changer le _baudrate_ sans redémarrer, il l'a utilisée.
Je pense que j'introduirai d'autres modification dans mon fork, mais ça va demander un peu  de travail. :)

**Comment by andromeda2911** [comment]
Sur pico ca ne compil pas.

`error: 'class arduino::HardwareSerial' has no member named 'setTX
`
Par contre sur pico Serial2.setRX fonctionne.
Il n'aime pas le passage en paramètre d'une fonction, HardwareSerial.

**Comment by andromeda2911** [comment]
Pour pico ca ne fonctionnera pas d'apres une recherche sur la meme erreur, il faudra faire une version pour pico, sur pico ce n'est pas HarwareSerial mais SerialUART.

`Minitel(SerialUART& serial, int8_t rx, int8_t tx);
`
Ou sinon mettre des directives de compilation #if defined etc .. idem pour le membre mySerial.

**Comment by troisiemetype** [comment]
Je n'avais pas testé. En effet, ça ne fonctionne pas avec le pico. La classe HardwareSerial est surchargée en SerialUART.
J'ai modifié ce que j'avais fait, maintenant ça compile.

**Comment by andromeda2911** [comment]
maintenant avec SerialUART ca fonctionne sur pico, mais j'ai préféré faire une librarie dédié au pico.
